### PR TITLE
Ruby - Bella A - C19

### PR DESCRIPTION
**Pull request by bellaiam**

### CODE REVIEW

File: adagrams/game.py [hunk 1]
```diff
@@ -1,11 +1,126 @@
...
+    #         score += 4
+    #     elif letter == "K":
+    #         score += 5
+    #     elif letter in "JX":
+    #         score += 8
+    #     elif letter in "QZ":
+    #         score += 10
+    # if len(word) >= 7:
+    #     score += 8
+    # return score
```
Comment (mmcknett-ada) [inline adagrams/game.py]: [nit] You can remove commented-out code. `git` remembers it!

File: adagrams/game.py [hunk 2]
```diff
@@ -1,11 +1,126 @@
...
+    scores = {
+        'A': 1, 'B': 3, 'C': 3, 'D': 2, 'E': 1, 'F': 4, 'G': 2, 'H': 4,
+        'I': 1, 'J': 8, 'K': 5, 'L': 1, 'M': 3, 'N': 1, 'O': 1, 'P': 3,
+        'Q': 10, 'R': 1, 'S': 1, 'T': 1, 'U': 1, 'V': 4, 'W': 4, 'X': 8,
+        'Y': 4, 'Z': 10}
```
Comment (mmcknett-ada) [inline adagrams/game.py]: [nit] It's easier to read this if it's split across multiple lines, similar to the way `LETTER_POOL` is above.

File: adagrams/game.py [hunk 3]
```diff
@@ -1,11 +1,126 @@
...
+    score = sum(scores.get(letter, 0) for letter in word.upper())
```
Comment (mmcknett-ada) [inline adagrams/game.py]: [suggestion] You might prefer to use `scores[letter]` instead of `scores.get(letter, 0)`. If you ever get a letter that isn't recognized, your program will crash with an error message, as opposed to simply not counting the letter. It can be easier to debug a crash than a subtle difference in score.

File: adagrams/game.py (hunk 1 above)
Comment (mmcknett-ada) [reply adagrams/game.py]: Same goes for `get_highest_word_score` and throughout the file.

File: adagrams/game.py [hunk 4]
```diff
@@ -1,11 +1,126 @@
...
+    'V': 2, 
+    'W': 2, 
+    'X': 1, 
+    'Y': 2, 
+    'Z': 1
+}
+
 def draw_letters():
-    pass
+    hand = random.sample(list(LETTER_POOL.keys()), 10)
```
Comment (mmcknett-ada) [inline adagrams/game.py]: This isn't *quite* right. Using `random.sample` this way makes it so that you never draw a letter more than once! To see this in action, you can add this test to the end of your wave 1 tests:

```python
def test_ever_draws_duplicate_letters():
    def has_dupes(letter_bank):
        unique_letters = set(letter_bank)
        return len(unique_letters) != len(letter_bank)

    saw_a_duplicate = False
    for i in range(1000):
        letters = draw_letters()
        saw_a_duplicate = saw_a_duplicate or has_dupes(letters)

    assert saw_a_duplicate
```

You want the likelihood of drawing any given letter to be based on the number of letters in the pool. Think of it like drawing ten letters out of a bag (the way you do with games like "bananagrams" or "scrabble") into your hand. When you sample from the list of `LETTER_POOL` keys, it's as though you only ever put each letter into the bag once! To do what you want with `random.sample`, you need to provide `counts`:

```python
hand = random.sample(list(LETTER_POOL.keys()), counts = list(LETTER_POOL.values()), k = 10)
```

Regarding built-in functions: when bringing in functions from imported modules (that we haven’t covered in the curriculum), please consider including a comment about how you understand the function to work, and how you researched it. Now that we’ve discussed Big O, it’s also useful to think about what the time/space complexity of the function might be. Thinking about how we can implement this functionality ourselves can help us gain insight into what that might be.

Remember that not every language has the same built-in functions, so we should be able to implement this same functionality ourselves. In the projects, we don’t ask for any features that we don’t think you should be able to write yourself. For drawing a random hand, the only external function that would be “required” is a way to pick a random number (such as randint). At this stage in our coding journeys, it’s often more valuable to re-implement things by hand to practice thinking about how to break down and approach the problem than to use a single library function to solve the problem for us.

File: adagrams/game.py [hunk 5]
```diff
@@ -1,11 +1,126 @@
...
+#         if letter not in letter_bank or bank_counts[letter] == 0:
+#             return False
+#         # Otherwise, decrement the count of the letter in bank_counts
+#         bank_counts[letter] -= 1
+#     # If we make it through the loop without returning False, return True
+#     return True
 def uses_available_letters(word, letter_bank):
-    pass
+    word = word.upper()
+    word_counts = Counter(word)
```
Comment (mmcknett-ada) [inline adagrams/game.py]: Great use of a built-in! For `Counter` as with the `sample` function above, please consider including a comment about how you understand the function to work, and how you researched it.

File: adagrams/game.py [hunk 6]
```diff
@@ -1,11 +1,126 @@
...
+    for letter, count in word_counts.items():
+        if count > letter_bank.count(letter):
```
Comment (mmcknett-ada) [inline adagrams/game.py]: I'm surprised you used `Counter` for `word` but not for `letter_bank`.

File: adagrams/game.py [hunk 7]
```diff
@@ -1,11 +1,126 @@
...
+#                 max_word = word
+#     return max_word, max_val
 def get_highest_word_score(word_list):
-    pass
\ No newline at end of file
+    # Calculate the score of each word in the list
+    word_scores = [(word, score_word(word)) for word in word_list]
+
+    # Find the maximum score among the list of words
+    max_score = max(word_scores, key=lambda x: x[1])[1]
```
Comment (mmcknett-ada) [inline adagrams/game.py:158]: [nit] One edge case came to mind: what happens if `word_list` is empty?

**Review by mmcknett-ada — COMMENTED** [review]
Looks good overall, but `draw_letters` has an issue the tests don't catch. Yellow 🟡 

Your tests are passing, and your code is concise and readable. I have some nitpicks, comments, and suggestions, but those don't detract from the fact that this is a solid submission.

Overall:
- Be careful with built-ins and knowing what they do. Specifically, have a look at my note on `random.sample`.
- You can remove commented-out code.
- I love how you used comments in `get_highest_word_score` to explain what was happening! They helped me read the code.
- Great job making your code "Pythonic".

File: adagrams/game.py (hunk 7 above)
Comment (bellaiam) [reply adagrams/game.py:158]: raised valueError

File: adagrams/game.py [hunk 8]
```diff
@@ -1,11 +1,175 @@
...
+#         # If the letter count is less than its value in LETTER_POOL, add it to the hand
+#         if letter_counts[letter] < LETTER_POOL[letter]:
+#             hand.append(letter)
+#             letter_counts[letter] += 1
+#         # Otherwise, try again
+#         else:
+#             i -= 1
+    
+#     # Return the list of drawn letters
+#     return hand
```
Comment (mmcknett-ada) [inline adagrams/game.py:82]: This one occasionally only returns a `hand` of 9 letters! See if you can figure out why.

To prevent this sort of issue, you can loop *while* the hand isn't big enough, e.g.:

```python
    while len(hand) < 10:
        letter = random.choice(list(LETTER_POOL.keys()))
        if letter_counts[letter] < LETTER_POOL[letter]:
            hand.append(letter)
            letter_counts[letter] += 1
```

**Comment by mmcknett-ada** [comment]
Great revisions! Green 🟢
